### ISSUE TITLE
WebDav export produced errors during git-annex's "unexport" stage. DAV failure: 404 "Not Found", HTTP request: "DELETE".

### ISSUE THREAD

**Issue by mslw** [datalad/datalad-next#202]
This is something where I couldn't figure whether it's an error caused by the remote server acting weirdly, caused by the dataset state, or coming from DataLad, so I'm copying all the information I have.

- I am pushing to a WebDAV sibling in export tree mode.
- The WebDAV sibling is on a Dracoon instance.
- I am pushing from a temporary clone to a location that has been pushed into before (including previous failed attempts), so there is a chance I brought some of the problems on myself due to the sibling's troubled history. But the files in the sibling look all right.
- I am working with the Python API and using `ds.push(to='dracoon', force='export')` (force export because push is from temporary clone, and DataLad would prevent pushing due to UUID mismatch). I also tried `force=all` with no change.
- DataLad 0.17.10 (actually maint branch from just before release), git-annex 10.20220504

The first time, I saw the following error (first line is my custom log entry, the rest is traceback):

```
2022-12-13 14:47:04,388 - run_updates - ERROR: Exception occurred on export to webdav
Traceback (most recent call last):
  File "workflow/run_updates.py", line 121, in <module>
    ds.push(to='dracoon', force='export')
  File "/data/group/psyinf/sfb1451/motor-assessment/venv-ma/lib/python3.7/site-packages/datalad/distribution/dataset.py", line 502, in apply_func
    return f(*args, **kwargs)
  File "/data/group/psyinf/sfb1451/motor-assessment/venv-ma/lib/python3.7/site-packages/datalad/interface/utils.py", line 340, in eval_func
    return return_func(*args, **kwargs)
  File "/data/group/psyinf/sfb1451/motor-assessment/venv-ma/lib/python3.7/site-packages/datalad/interface/utils.py", line 330, in return_func
    results = list(results)
  File "/data/group/psyinf/sfb1451/motor-assessment/venv-ma/lib/python3.7/site-packages/datalad/interface/utils.py", line 509, in _execute_command_
    msg="Command did not complete successfully")
datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 3 failed:
[{'action': 'copy',
  'error_message': 'DAV failure: Status {statusCode = 404, statusMessage = '
                   '"Not Found"} "" HTTP request: "DELETE" '
                   '"/webdav/Neuro-CRC1451/Z03/motor-export/code"',
  'path': '/data/group/psyinf/sfb1451/motor-assessment/motor-assessment-derived/code/run_updates.py',
  'refds': '/data/group/psyinf/sfb1451/motor-assessment/motor-assessment-derived',
  'status': 'error',
  'target': 'dracoon-storage',
  'type': 'dataset'},
 {'action': 'copy',
  'error_message': 'DAV failure: Status {statusCode = 404, statusMessage = '
                   '"Not Found"} "" HTTP request: "DELETE" '
                   '"/webdav/Neuro-CRC1451/Z03/motor-export/output"',
  'path': '/data/group/psyinf/sfb1451/motor-assessment/motor-assessment-derived/output/table.csv',
  'refds': '/data/group/psyinf/sfb1451/motor-assessment/motor-assessment-derived',
  'status': 'error',
  'target': 'dracoon-storage',
  'type': 'dataset'},
 {'action': 'publish',
  'hints': 'Updates were rejected because the remote contains work that you do '
           'not have locally. This is usually caused by another repository '
           'pushing to the same ref. You may want to first integrate the '
           "remote changes (e.g., 'git pull ...') before pushing again. See "
           "the 'Note about fast-forwards' in 'git push --help' for details.",
  'message': 'refs/heads/master->dracoon:refs/heads/master [rejected] (fetch '
             'first)',
  'operations': ['rejected', 'error'],
  'path': '/data/group/psyinf/sfb1451/motor-assessment/motor-assessment-derived',
  'refds': '/data/group/psyinf/sfb1451/motor-assessment/motor-assessment-derived',
  'refspec': 'refs/heads/master:refs/heads/master',
  'status': 'error',ere is
  'target': 'dracoon',
  'type': 'dataset'}]
```

Running this again, but looking into the result renderer during operation, instead of traceback:

```
copy(ok): .gitmodules (dataset)
copy(ok): code/update_output.py (dataset)
copy(ok): README.md (dataset)
copy(error): code/run_updates.py (dataset) [DAV failure: Status {statusCode = 404, statusMessage = "Not Found"} "" HTTP request: "DELETE" "/webdav/Neuro-CRC1451/Z03/motor-export/code"]
copy(ok): output/summary.txt (dataset)
copy(error): output/table.csv (dataset) [DAV failure: Status {statusCode = 404, statusMessage = "Not Found"} "" HTTP request: "DELETE" "/webdav/Neuro-CRC1451/Z03/motor-export/output"]
copy(ok): requirements.txt (dataset)
copy(ok): .datalad/.gitattributes (dataset)
copy(ok): .datalad/config (dataset)
copy(ok): .gitattributes (dataset)
copy(ok): .gitignore (dataset)
copy(ok): .gitmodules (dataset)
copy(ok): README.md (dataset)
copy(ok): code/.gitattributes (dataset)
copy(ok): code/.gitignore (dataset)
copy(ok): code/create_csv_table.py (dataset)
  [9 similar messages have been suppressed; disable with datalad.ui.suppress-similar-results=off]
publish(ok): . (dataset) [refs/heads/git-annex->dracoon:refs/heads/git-annex 3714b67..07ff87d]
publish(error): . (dataset) [refs/heads/master->dracoon:refs/heads/master [rejected] (fetch first)]
action summary:                                                                                                                        
  copy (error: 2, ok: 23)
  publish (error: 1, ok: 1)
Hints: 
1: Updates were rejected because the remote contains work that you do not have locally. This is usually caused by another repository pushing to the same ref. You may want to first integrate the remote changes (e.g., 'git pull ...') before pushing again. See the 'Note about fast-forwards' in 'git push --help' for details.
```

Running the `git annex export` command:
```
❱ git annex export HEAD --to dracoon-storage --json --json-error-messages
{"command":"unexport dracoon-storage","success":true,"input":[],"error-messages":[],"file":".gitmodules"}
{"command":"unexport dracoon-storage","success":true,"input":[],"error-messages":[],"file":"code/update_output.py"}
{"command":"unexport dracoon-storage","success":true,"input":[],"error-messages":[],"file":"README.md"}
{"command":"unexport dracoon-storage","success":false,"input":[],"error-messages":["  DAV failure: Status {statusCode = 404, statusMessage = \"Not Found\"} \"\" HTTP request: \"DELETE\" \"/webdav/Neuro-CRC1451/Z03/motor-export/code\""],"file":"code/run_updates.py"}
{"command":"unexport dracoon-storage","success":true,"input":[],"error-messages":[],"file":"output/summary.txt"}
{"command":"unexport dracoon-storage","success":false,"input":[],"error-messages":["  DAV failure: Status {statusCode = 404, statusMessage = \"Not Found\"} \"\" HTTP request: \"DELETE\" \"/webdav/Neuro-CRC1451/Z03/motor-export/output\""],"file":"output/table.csv"}
{"command":"unexport dracoon-storage","success":true,"input":[],"error-messages":[],"file":"requirements.txt"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":".datalad/.gitattributes"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":".datalad/config"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":".gitattributes"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":".gitignore"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":".gitmodules"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"README.md"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"code/.gitattributes"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"code/.gitignore"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"code/create_csv_table.py"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"code/run_updates.py"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"code/update_output.py"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"output/cleaned.csv"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"output/reports/duplicates.csv"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"output/reports/incorrectly-signed.csv"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"output/summary.txt"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"output/table.csv"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"requirements.txt"}
{"command":"export dracoon-storage","success":true,"input":[],"error-messages":[],"file":"setup.zsh"}
```

So this is an 404 response to a delete export during "unexport". Which is followed by a successful "export", and the file is placed into the WebDAV location. But the file which 404'd was there before starting this export!

Running the same `git annex export` command afterwards (at this point we know 100% that the files are there, because their export succeeded and I saw them in web UI) gives 404 during "unexport" once again *for the same two files* (and one more, which was first exported by doing the step above). All other files unexport and export just fine:

```
{"command":"unexport dracoon-storage","success":false,"input":[],"error-messages":["  DAV failure: Status {statusCode = 404, statusMessage = \"Not Found\"} \"\" HTTP request: \"DELETE\" \"/webdav/Neuro-CRC1451/Z03/motor-export/code\""],"file":"code/run_updates.py"}
{"command":"unexport dracoon-storage","success":true,"input":[],"error-messages":[],"file":"code/update_output.py"}
{"command":"unexport dracoon-storage","success":false,"input":[],"error-messages":["  DAV failure: Status {statusCode = 404, statusMessage = \"Not Found\"} \"\" HTTP request: \"DELETE\" \"/webdav/Neuro-CRC1451/Z03/motor-export/output\""],"file":"output/cleaned.csv"}
{"command":"unexport dracoon-storage","success":false,"input":[],"error-messages":["  DAV failure: Status {statusCode = 404, statusMessage = \"Not Found\"} \"\" HTTP request: \"DELETE\" \"/webdav/Neuro-CRC1451/Z03/motor-export/output/reports\""],"file":"output/reports/duplicates.csv"}
```

I also tried `fsck`:

```
❱ git annex fsck --fast --from dracoon-storage 
fsck output/cleaned.csv 
git-annex: Conflict detected. Different trees have been exported to dracoon-storage. Use git-annex export to resolve this conflict.
ok
fsck output/reports/duplicates.csv ok
fsck output/reports/incorrectly-signed.csv ok
fsck output/summary.txt ok
fsck output/table.csv ok
```

But after fsck, same story repeats - neither `fsck` nor `export` seem to resolve the conflict.

In the end, I addad a `try ... except IncompleteResultsError` to my code to ignore the (un)export errors, as all the files at the remote end look just fine.

But I can't say whether the 404 errors are just wrong behaviour of the server, consequence of my error leading to weird states in annex history, or something else. Consequently, I can't tell how they should be treated (git annex does say, "success":false). 

tl; dr
- There is `copy(error): output/table.csv (dataset) [DAV failure: Status {statusCode = 404, statusMessage = "Not Found"} "" HTTP request: "DELETE" "/webdav/Neuro-CRC1451/Z03/motor-export/output"]` from DataLad
- The underlying git-annex output comes from "unexport" and is `{"command":"unexport dracoon-storage","success":false,"input":[],"error-messages":["  DAV failure: Status {statusCode = 404, statusMessage = \"Not Found\"} \"\" HTTP request: \"DELETE\" \"/webdav/Neuro-CRC1451/Z03/motor-export/output\""],"file":"output/table.csv"}`
- Nevertheless, git-annex then attempts "export" on the file and succeeds
- The same failure repeats on subsequent exports

**Comment by mslw** [comment]
I stopped using that server for unrelated reasons, so I think the issue can be closed and kept archived, as it seems unlikely that we will get a chance to investigate it further.

My initial conclusion remains valid:
> I can't say whether the 404 errors are just wrong behaviour of the server, consequence of my error leading to weird states in annex history, or something else
